src/Hazelcast.Net/Networking/AddressProvider.cs

```diff
@@ -85,20 +85,36 @@ public AddressProvider(NetworkingOptions networkingOptions, ILoggerFactory logge
             }
         }
 
+        // internal constructor for tests (so we can test with a bogus create map method)
+        internal AddressProvider(NetworkingOptions networkingOptions, Func<IDictionary<NetworkAddress, NetworkAddress>> createMap, bool hasMap, ILoggerFactory loggerFactory)
```

Comment:
You are putting extra code into the production code even if it is marked as `internal` which is only needed for testing (I understand that it can not be used by the lib users). Will it not get placed in the produced used dll? Is there no other way such as stubbing a class in .NET you can use and not add extra code to production?

Reply:
No. If that class is to have that constructor, then... it needs to be here and then yes, the code goes in the production DLL. I don't know how to test otherwise.

Reply:
It is usually done via `mock` libraries. I believe there are mock frameworks in .NET, at least when I searched for it on the internet, e.g. [this](https://stackoverflow.com/questions/20400734/how-do-i-mock-a-class-without-an-interface) one. I will question this approach in the team to check how other clients are. I believe it is important not to ship test code for a production library but let's get some ideas on this.

Reply:
I think this code already does mocking. The tests can "mock" the createMap interface by passing a custom implementation via this ctor. But instead of using interfaces/virtual functions, it is using a function pointer (or whatever .net calls them).

I think the way it works is a bit confusing because the AddressProvider provides a default implementation if you don't want to mock it. Something like this could be clearer:

```
class AddressProvider {
    // have just one ctor:
    AddressProvider(IList<string> addresses, Func<...> createMap);
}

// put the default createMap method somewhere outside, and provide it to AddressProvider during construction
IDictionary<...> defaultCreateMap(cloudConfig) {
   // implement the "if cloud enabled ... else ..." logic here
}

// in tests
IDictionary<...> myMockCreateMap() {
  return {"bogus": "mapping"};
}

var p = new AddressProvider({}, myMockCreateMap)
```

Reply:
Well, one way or another, the code is there in the production release. We want to be able to inject a `createMap` implementation, we need a constructor to support it. Now we can refactor the existing way to use that constructor and then say, see, it's code that is used, but apart from making things more complex I fail to see how it can help.

I also fail to see how mocking would change anything here. In fact, it is precisely what we are doing: supporting injecting a `createMap` method, including a bogus one. What that new constructor does is, precisely, support mocking.

One way to deal with this is to `#ifdef` the code, and have a "production release" build that is different from a "test release" build - but that sounds quite error prone to me. It means that the production binary we release is not what we test.

At that point, I fail to see how I can test what we are trying to test, without adding that ctor. Ideas welcomed.

Reply:
Might have an idea actually. Stay tuned.

Reply:
There, change pushed, now the `AddressProvider` relies on a source, which can be mocked in the test project. So, no more "dead" code in the production dll. Let me know what you think, it's probably cleaner.

Reply:
> Well, one way or another, the code is there in the production release. We want to be able to inject a createMap implementation, we need a constructor to support it. Now we can refactor the existing way to use that constructor and then say, see, it's code that is used, but apart from making things more complex I fail to see how it can help.

I agree that it makes no practical difference (you still have the code but now it is "used") and I was completely fine with the way it previously was. I was only trying to suggest a compromise that would satisfy both concerns: no dead code + testable via mocking. 

I think everyone should be perfectly fine with the current version.

Reply:
Hey, let me know what you think of the latest changes.